File: src/lac/bicgstab.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const assert = std.debug.assert;

const lac = @import("lac.zig");

const IdentityMap = lac.IdentityMap;
const isLinearMap = lac.isLinearMap;
const hasLinearMapCallback = lac.hasLinearMapCallback;
const isLinearSolver = lac.isLinearSolver;

pub const BiCGStabSolver = struct {
    max_iters: usize,
    tolerance: f64,

    // Alias
    const Self = @This();

    // Constructs a new bicgstab solver
    pub fn new(max_iters: usize, tolerance: f64) Self {
        return .{
            .max_iters = max_iters,
            .tolerance = tolerance,
        };
    }

    pub const Error = error{OutOfMemory};

    pub fn solve(self: *const Self, allocator: Allocator, oper: anytype, x: []f64, b: []const f64) Error!void {
        assert(x.len == b.len);

        // Compute tolerance
        const tol = @fabs(self.tolerance) * norm(b);

        if (norm(b) <= 1e-60) {
            @memset(x, 0.0);
            return;
        }

        // Allocate scratch vectors
        const ndofs = x.len;

        const rg: []f64 = try allocator.alloc(f64, ndofs);
        defer allocator.free(rg);

        const rh: []f64 = try allocator.alloc(f64, ndofs);
        defer allocator.free(rh);

        const pg: []f64 = try allocator.alloc(f64, ndofs);
        defer allocator.free(pg);

        const ph: []f64 = try allocator.alloc(f64, ndofs);
        defer allocator.free(ph);

        const sg: []f64 = try allocator.alloc(f64, ndofs);
        defer allocator.free(sg);

        const sh: []f64 = try allocator.alloc(f64, ndofs);
        defer allocator.free(sh);

        const tg: []f64 = try allocator.alloc(f64, ndofs);
        defer allocator.free(tg);

        const vg: []f64 = try allocator.alloc(f64, ndofs);
        defer allocator.free(vg);

        const tp: []f64 = try allocator.alloc(f64, ndofs);
        defer allocator.free(tp);

        // Set termination tolerance
        oper.apply(tp, x);

        for (0..ndofs) |i| {
            rg[i] = b[i] - tp[i];
        }

        @memcpy(rh, rg);
        @memset(sh, 0.0);
        @memset(ph, 0.0);

        var residual = norm(rg);

        var iter: usize = 0;
        var rho0: f64 = 0.0;
        var rho1: f64 = 0.0;
        var pra: f64 = 0.0;
        var prb: f64 = 0.0;
        var prc: f64 = 0.0;

        while (iter < self.max_iters) : (iter += 1) {
            rho1 = dot(rg, rh);

            if (rho1 == 0.0) {
                break;
            }

            if (iter == 0) {
                @memcpy(pg, rg);
            } else {
                prb = (rho1 * pra) / (rho0 * prc);

                for (0..ndofs) |i| {
                    pg[i] = rg[i] + prb * (pg[i] - prc * vg[i]);
                }
            }

            rho0 = rho1;

            // Identity PC
            @memcpy(ph, pg);
            oper.apply(vg, ph);

            pra = rho1 / dot(rh, vg);

            for (0..ndofs) |i| {
                sg[i] = rg[i] - pra * vg[i];
            }

            if (norm(sg) <= 1e-60) {
                for (0..ndofs) |i| {
                    x[i] = x[i] + pra * ph[i];
                }

                oper.apply(tp, x);

                for (0..ndofs) |i| {
                    rg[i] = b[i] - tp[i];
                }

                residual = norm(rg);

                break;
            }

            @memcpy(sh, sg);
            oper.apply(tg, sh);

            prc = dot(tg, sg) / dot(tg, tg);
            for (0..ndofs) |i| {
                x[i] = x[i] + pra * ph[i] + prc * sh[i];
                rg[i] = sg[i] - prc * tg[i];
            }

            residual = norm(rg);

            if (comptime hasLinearMapCallback(@TypeOf(oper))) {
                oper.callback(iter, residual, x);
            }

            if (residual <= tol) break;
        }

        if (iter < self.max_iters) iter += 1;
    }

    fn norm(slice: []const f64) f64 {
        return @sqrt(dot(slice, slice));
    }

    fn dot(u: []const f64, v: []const f64) f64 {
        var result: f64 = 0.0;
        for (u, v) |a, b| {
            result += a * b;
        }
        return result;
    }

    fn axpby(a: f64, x: []const f64, b: f64, y: []f64) void {
        for (x, y) |xv, *yv| {
            yv.* = xv * a + yv.* * b;
        }
    }
};

test "BiCGStab convergence" {
    const expect = std.testing.expect;
    const expectEqualSlices = std.testing.expectEqualSlices;

    try expect(isLinearSolver(BiCGStabSolver));

    const allocator = std.testing.allocator;
    const ndofs = 100;

    const x = try allocator.alloc(f64, ndofs);
    defer allocator.free(x);

    const b = try allocator.alloc(f64, ndofs);
    defer allocator.free(b);

    for (0..ndofs) |i| {
        x[i] = 0.0;
        b[i] = @floatFromInt(i);
    }

    const solver = BiCGStabSolver.new(1000, 1e-10);

    try solver.solve(allocator, IdentityMap{}, x, b);

    try expectEqualSlices(f64, b, x);
}

File: src/lac/bicgstab.rs
#![allow(clippy::needless_range_loop)]

use super::{LinearMap, LinearSolver};

/// Implementation of the stabilized bi-conjugate gradient method.
pub struct BiCGStabSolver {
    max_iterations: usize,
    tolerance: f64,
    dimension: usize,

    rg: Vec<f64>,
    rh: Vec<f64>,
    pg: Vec<f64>,
    ph: Vec<f64>,
    sg: Vec<f64>,
    sh: Vec<f64>,
    tg: Vec<f64>,
    vg: Vec<f64>,
    tp: Vec<f64>,

    mask: Vec<bool>,
}

impl BiCGStabSolver {
    /// Retrieves the tolerance of the multigrid solver.
    pub fn tolerance(&self) -> f64 {
        self.tolerance
    }

    /// Retrieves the maximum number of iterations before the multigrid
    /// solver fails.
    pub fn max_iterations(&self) -> usize {
        self.max_iterations
    }
}

#[derive(Debug)]
pub enum BiCGStabError {
    Breakdown,
    FailedToConverge,
}

pub struct BiCGStabConfig {
    pub max_iterations: usize,
    pub tolerance: f64,
}

impl LinearSolver for BiCGStabSolver {
    type Error = BiCGStabError;
    type Config = BiCGStabConfig;

    /// Builds a new BiCGStabSolver with the given dimension and
    /// configuration.
    fn new(dimension: usize, config: &BiCGStabConfig) -> Self {
        Self {
            dimension,
            max_iterations: config.max_iterations,
            tolerance: config.tolerance,

            rg: vec![0.0; dimension],
            rh: vec![0.0; dimension],
            pg: vec![0.0; dimension],
            ph: vec![0.0; dimension],
            sg: vec![0.0; dimension],
            sh: vec![0.0; dimension],
            tg: vec![0.0; dimension],
            vg: vec![0.0; dimension],
            tp: vec![0.0; dimension],

            mask: vec![true; dimension],
        }
    }

    fn dimension(&self) -> usize {
        self.dimension
    }

    fn solve<M: LinearMap>(
        &mut self,
        mut map: M,
        rhs: &[f64],
        solution: &mut [f64],
    ) -> Result<(), Self::Error> {
        // Mask dofs
        map.mask(&mut self.mask);
        // Mask closure
        let mask = |dest: &mut [f64]| {
            for i in
                dest.iter_mut()
                    .zip(self.mask.iter())
                    .filter_map(|(v, &m)| if !m { Some(v) } else { None })
            {
                *i = 0.0;
            }
        };

        // Set termination tolerance
        map.apply(solution, &mut self.tp);
        mask(&mut self.tp);

        // let nrhs = norm(rhs);

        for i in 0..self.dimension {
            self.rg[i] = rhs[i] - self.tp[i];
        }
        mask(&mut self.rg);

        self.rh.clone_from_slice(&self.rg);
        self.sh.fill(0.0);
        self.ph.fill(0.0);

        let mut residual = norm(&self.rg);

        let tolerance = self.tolerance * residual;

        let mut iter = 0;
        let mut rho0 = 0.0;
        let mut rho1 = 0.0;
        let mut pra = 0.0;
        let mut prb = 0.0;
        let mut prc = 0.0;

        _ = residual;
        _ = rho1;
        _ = prb;

        while iter < self.max_iterations {
            rho1 = dot(&self.rg, &self.rh);

            if rho1 == 0.0 {
                return Err(BiCGStabError::Breakdown);
            }

            if iter == 0 {
                self.pg.clone_from_slice(&self.rg);
            } else {
                prb = (rho1 * pra) / (rho0 * prc);

                for i in 0..self.dimension {
                    self.pg[i] = self.rg[i] + prb * (self.pg[i] - prc * self.vg[i]);
                }
                mask(&mut self.pg);
            }

            rho0 = rho1;

            // Identity Preconditioner
            self.ph.clone_from_slice(&self.pg);
            map.apply(&self.ph, &mut self.vg);

            pra = rho1 / dot(&self.rh, &self.vg);

            // if dot(&self.rh, &self.vg).is_nan() {
            //     println!("DOT Hmm");
            //     println!("RH {:?}", &self.rh);
            //     println!("VG {:?}", &self.vg);

            //     panic!("Testing");
            // }

            for i in 0..self.dimension {
                self.sg[i] = self.rg[i] - pra * self.vg[i];
            }

            mask(&mut self.sg);

            if norm(&self.sg) <= 1e-60 {
                for i in 0..self.dimension {
                    solution[i] += pra * self.ph[i];
                }

                map.apply(solution, &mut self.tp);

                for i in 0..self.dimension {
                    self.rg[i] = rhs[i] - self.tp[i];
                }

                mask(&mut self.rg);

                // residual = norm(&self.rg);
                break;
            }

            self.sh.clone_from_slice(&self.sg);
            map.apply(&self.sh, &mut self.tg);

            prc = dot(&self.tg, &self.sg) / dot(&self.tg, &self.tg);

            for i in 0..self.dimension {
                solution[i] = solution[i] + pra * self.ph[i] + prc * self.sh[i];
                self.rg[i] = self.sg[i] - prc * self.tg[i];
            }

            mask(solution);
            mask(&mut self.rg);

            residual = norm(&self.rg);
            map.callback(iter, residual, solution);

            if residual <= tolerance {
                break;
            }

            iter += 1;
        }

        if iter < self.max_iterations {
            iter += 1;
        }

        if iter >= self.max_iterations {
            return Err(BiCGStabError::FailedToConverge);
        }

        Ok(())
    }
}

fn dot(a: &[f64], b: &[f64]) -> f64 {
    a.iter().zip(b.iter()).map(|(a, b)| a * b).sum()
}

fn norm(a: &[f64]) -> f64 {
    dot(a, a).sqrt()
}

#[cfg(test)]
mod tests {
    use super::{BiCGStabConfig, BiCGStabError, BiCGStabSolver};
    use crate::lac::{IdentityMap, LinearSolver};

    #[test]
    fn identity_convergence() -> Result<(), BiCGStabError> {
        let mut solution = vec![0.0; 100];
        let mut rhs = vec![0.0; 100];

        for i in 0..100 {
            solution[i] = 0.0;
            rhs[i] = i as f64;
        }

        let mut solver: BiCGStabSolver = BiCGStabSolver::new(
            100,
            &BiCGStabConfig {
                max_iterations: 1000,
                tolerance: 10e-12,
            },
        );

        solver.solve(IdentityMap::new(100), &rhs, &mut solution)?;

        assert_eq!(solution == rhs, true);

        Ok(())
    }
}
